pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -106,182 +180,237 @@ func (r *reconciler) Reconcile(rule *CompletedRule) error {
-			proto := corev1.Protocol(protoStr)
-			service.Protocol = &proto
-		}
-		if s.Port != nil {
-			// Ignore named port for now.
-			if s.Port.Type == intstr.String {
-				continue
-			}
-			service.Port = s.Port
+	for svcHash, ofRule := range ofRuleByServicesMap {
+		ofID, err := r.installOFRule(ofRule)
+		if err != nil {
+			return err
 		}
-		out = append(out, service)
+		// Record ofID only if its Openflow is installed successfully.
+		lastRealized.ofIDs[svcHash] = ofID
 	}
-	return out
+
+	return nil
 }
 
 // update calculates the difference of Addresses between oldRule and newRule,
 // and invokes Openflow client's methods to reconcile them.
 func (r *reconciler) update(lastRealized *lastRealized, newRule *CompletedRule) error {
 	klog.V(2).Infof("Updating existing rule %v", newRule)
+	// staleOFIDs tracks servicesHash that are no long needed.
+	// Firstly fill it with the last realized ofIDs.
+	staleOFIDs := make(map[servicesHash]uint32, len(lastRealized.ofIDs))
+	for svcHash, ofID := range lastRealized.ofIDs {
+		staleOFIDs[svcHash] = ofID
+	}
+
 	// As rule identifier is calculated from the rule's content, the update can
 	// only happen to Group members.
-	var addedFrom, addedTo, deletedFrom, deletedTo []types.Address
-	var newOFPorts sets.Int32
-	var newIPs sets.String
 	if newRule.Direction == v1beta1.DirectionIn {
-		for _, a := range newRule.FromAddresses.Difference(lastRealized.FromAddresses) {
-			addedFrom = append(addedFrom, ipAddressToOFAddress(a.IP))
-		}
-		for _, a := range lastRealized.FromAddresses.Difference(newRule.FromAddresses) {
-			deletedFrom = append(deletedFrom, ipAddressToOFAddress(a.IP))
-		}
-		newOFPorts = r.podsToOFPorts(newRule.Pods)
-		for p := range newOFPorts.Difference(lastRealized.podOFPorts) {
-			addedTo = append(addedTo, openflow.NewOFPortAddress(p))
-		}
-		for p := range lastRealized.podOFPorts.Difference(newOFPorts) {
-			deletedTo = append(deletedTo, openflow.NewOFPortAddress(p))
+		from1 := podsToOFAddresses(newRule.FromAddresses)
+		from2, exceptFrom := ipBlocksToOFAddresses(newRule.From.IPBlocks)
+		addedFrom := podsToOFAddresses(newRule.FromAddresses.Difference(lastRealized.FromAddresses))
+		deletedFrom := podsToOFAddresses(lastRealized.FromAddresses.Difference(newRule.FromAddresses))
+
+		podsByServicesMap, servicesMap := groupPodsByServices(newRule.Services, newRule.Pods)
+		for svcHash, pods := range podsByServicesMap {
+			newOFPorts := r.getPodOFPorts(pods)
+			ofID, exists := lastRealized.ofIDs[svcHash]
+			// Install a new Openflow rule if this group doesn't exist, otherwise do incremental update.
+			if !exists {
+				ofRule := &types.PolicyRule{
+					Direction:  networkingv1.PolicyTypeIngress,
+					From:       append(from1, from2...),
+					ExceptFrom: exceptFrom,
+					To:         ofPortsToOFAddresses(newOFPorts),
+					Service:    servicesToNetworkPolicyPort(servicesMap[svcHash]),
+				}
+				ofID, err := r.installOFRule(ofRule)
+				if err != nil {
+					return err
```

Comment:
do we want to return here or continue processing other pods and serviceS?

Reply:
There could be pros and cons in either way, even continue processing, it's not easy to decide what lastRealized should be for such intermediate case, and it should mean ovs's connection has problem so we are not likely to succeed for other pods too, so keep the processing simple terminate for now. It may need more fine-grained state (per group) tracking to  handle this situation, added TODO to improve it.